### PR TITLE
storage: prevent optimistic concurrency failures for patch

### PR DESCRIPTION
**Pull request by milas**
There's two classes of optimistic concurrency failures:
 1. Object update is outdated by time it's received by server
 2. Object is written to between update start and FS write

The first case should result in an immediate error propagated to
the user - there's no automatic resolution possible on the server.
(This is largely the optimistic concurrency check we've had thus
far.)

The second one is trickier: the first step at addressing this was
done in #53 (d5d681f238e029bfab17c270e3189c09d2d26497), which at
prevents two writes for the same version from succeeding. However,
this means that patch operations can fail due to optimistic
concurrency, which is explicitly against the apiserver contract.

As a result, this adds a "guaranteed update" method (name taken
from the etcd code that does logically the same thing). This will
retry on any FS/storage-level optimistic concurrency issues, by
replaying the whole update attempt from object fetch.

For an object replace, this means if it's valid at operation start
but another operation beats it before it's persisted, it'll retry,
fetch the new version of the object (or `nil` if object was deleted!)
and then trigger the logic to detect (1), as there will now be a
version mismatch _within the update logic_.

For an object patch, the patch will get replayed on the new object
version and persistence re-attempted. This will continue indefinitely
until a write actually succeeds or the context is canceled.

Additionally, to support this, object versioning has been overhauled
so that it's tracked per-object. For `MemoryFS`, it behaves like etcd
storage, where the versions are removed pre-serialization and restored
on read. Internally, the map now holds a tuple of the version and the
object data (eliminating the separate versions map to simplify book
keeping). This also enables an optimization to avoid incrementing the
version if the object hasn't changed at all. This is useful to avoid
unnecessary optimistic concurrency failures (client A submits no-op
update from v1, client B submits real update from v1, previously if
client B was processed after A, it would fail, now it'll still succeed
since A's no-op update will _really_ be a no-op.) This also matches
K8s apiserver/etcd behavior.

TL;DR Optimistic concurrency is respected throughout the entire
 update lifecycle AND patch works as expected!

### CODE REVIEW

**Review by nicks — COMMENTED** [review]
i feel like this needs some test that assert how we want concurrent updates to behave

e.g.,

if we kick off 20 goroutines that all write the same version, i would expect exactly one to succeed

File: pkg/storage/filepath/jsonfile_rest.go [hunk 1]
```diff
@@ -206,110 +201,125 @@ func (f *filepathREST) Update(
...
-
-	// TODO: should not be necessary, verify Get works before creating filepath
-	if f.NamespaceScoped() {
-		// ensures namespace dir
-		ns, ok := genericapirequest.NamespaceFrom(ctx)
-		if !ok {
-			return nil, false, ErrNamespaceNotExists
+	var isCreate bool
+	var isDelete bool
+	obj, err := f.guaranteedUpdate(ctx, name, func(input runtime.Object) (output runtime.Object, err error) {
```
[the remark this answers is not in the document]
Comment (milas) [reply pkg/storage/filepath/jsonfile_rest.go:206]: The name here comes from K8s/etcd storage code, I can add some comments (and I'm open to suggestions on a different name!)

There are two `UpdatedObjectInfo` variants:
* Replace (`PUT`) -> ignore `storageObj`, just return `newObj`
* Patch -> take `storageObj`, apply patch

Remember that patch is a partial object replace, i.e. specific fields, with last-one-wins behavior.

Thus, by retrying a patch on a _storage-level_ concurrency failure (a write went in since we fetched the obj to mutate but before we got to write it), we'll keep fetching the new version, applying our mutations, and attempting persistence as long as we get a version conflict _from the storage layer_ (`MemoryFS::Write`). We'll never fail the _handler layer_ (`filepathREST::Update`) version check because we've copied+mutated the storage object each time, so we're always "up-to-date" from the handler's perspective.

It's totally possible to hit this same scenario for a replace (obj was current at start of update but out of date by time we wrote it to storage)! But on the retry, the updated `storageObj` version will now no longer match our `newObj` version, so the _handler layer_ (`filepathREST::Update`) will return a conflict error at that point.

File: pkg/storage/filepath/version.go [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+import (
+	"fmt"
+	"strconv"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+func getResourceVersion(obj runtime.Object) (uint64, error) {
+	if obj == nil {
```
[the remark this answers is not in the document]
Comment (milas) [reply pkg/storage/filepath/version.go:12]: (FWIW `meta.Accessor()` will return an error if passed nil, so won't panic)

There is create-on-update logic where you end up getting with a `nil` object in the update paths, so this was convenience for that.

Sidenote: In the K8s/etcd code, there's a second arg for the `UpdateFunc` you pass in which is `ResponseMeta` that includes the revision + TTL from the server, so they use that instead of reading from the obj within the fn - maybe that's better? 🤷

File: pkg/storage/filepath/fs.go [hunk 3]
```diff
@@ -186,23 +202,41 @@ func (fs *MemoryFS) Write(encoder runtime.Encoder, p string, obj runtime.Object,
...
 	dir, err := fs.ensureDir(filepath.Dir(p))
 	if err != nil {
 		return err
 	}
 
-	newVersion, err := resourceVersion(obj)
-	if err != nil {
+	// increment the resource version - it's applied to the object pointer for
+	// the caller in addition to being used to ensure the write is valid
+	newVersion := storageVersion + 1
```
[the remark this answers is not in the document]
Comment (milas) [reply pkg/storage/filepath/fs.go]: Ahhh I misunderstood and thought etcd was doing per-key revisions!

Switched back to global in e2ae723dae341170bd3b1da300c9ca8aa02c9a8f (though it now lives in `MemoryFS`)

Note that `RealFS` objects still have a per-object incrementing revision [serialized with the obj] so that they'll behave sanely across app restarts

File: pkg/storage/filepath/fs.go [hunk 4]
```diff
@@ -186,23 +202,41 @@ func (fs *MemoryFS) Write(encoder runtime.Encoder, p string, obj runtime.Object,
...
+	if err := setResourceVersion(obj, newVersion); err != nil {
 		return err
 	}
 
-	dir[filepath.Base(p)] = buf
-	fs.versions[versionKey] = newVersion
+	dir[filepath.Base(p)] = versionedData{
+		version: newVersion,
+		data:    append(make([]byte, 0, len(buf.Bytes())), buf.Bytes()...),
```
[the remark this answers is not in the document]
Comment (milas) [reply pkg/storage/filepath/fs.go:239]: Thought was to create a new slice of the actual data size since the underlying buffer array might be 2x bigger than needed based on how it grew during encoding.

(I can always swap this back if preferred, I just ended up opportunistically swapping to `[]byte` when adding `versionedData` struct.)

File: pkg/storage/filepath/version.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (milas) [reply pkg/storage/filepath/version.go:12]: (I can also explicitly handle the 2 `nil` special cases if you think this is a footgun)

File: pkg/storage/filepath/jsonfile_rest.go [hunk 5]
```diff
@@ -423,6 +437,62 @@ func (f *filepathREST) objectDirName(ctx context.Context) string {
...
+// account the current contents of the object when deciding how the update object
+// should look.
+//
+// The "guaranteed" in the name comes from a method of the same name in the
+// Kubernetes apiserver/etcd code. Most of this method comment is copied from
+// its godoc.
+//
+// See https://github.com/kubernetes/apiserver/blob/544b6014f353b0f5e7c6fd2d3e04a7810d0ba5fc/pkg/storage/interfaces.go#L205-L238
+func (f *filepathREST) guaranteedUpdate(ctx context.Context, name string, tryUpdate updateFunc) (runtime.Object, error) {
+	for {
```
Comment (nicks) [inline pkg/storage/filepath/jsonfile_rest.go:460]: i think it would be good to put some sort of limit on this infinite loop, in case of bugs

File: pkg/storage/filepath/fs.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (nicks) [reply pkg/storage/filepath/fs.go:239]: hmmm...feels like an unsubstantiated optimization. i've also been wondering if we should use DeepCopyObject() rather than storing as bytes, to save on heap size.

**Review by nicks — APPROVED** [review]

File: pkg/storage/filepath/fs.go [hunk 6]
```diff
@@ -186,23 +204,41 @@ func (fs *MemoryFS) Write(encoder runtime.Encoder, p string, obj runtime.Object,
...
 	dir, err := fs.ensureDir(filepath.Dir(p))
 	if err != nil {
 		return err
 	}
 
-	newVersion, err := resourceVersion(obj)
-	if err != nil {
+	// increment the resource version - it's applied to the object pointer for
+	// the caller in addition to being used to ensure the write is valid
+	newVersion := atomic.AddUint64(&fs.rev, 1)
```
Comment (nicks) [inline pkg/storage/filepath/fs.go:232]: i don't think you need an atomic here? you already have the Lock

File: pkg/storage/filepath/fs.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (milas) [reply pkg/storage/filepath/fs.go]: I have no idea what's going on with GItHub review comments (I can't respond to the other comment re: this for some reason...?) but the buffer copy is gone as of b3db9ea

File: pkg/storage/filepath/jsonfile_rest.go (hunk 5 above)
Comment (milas) [reply pkg/storage/filepath/jsonfile_rest.go]: good call - capped it in 474a594 - it'll return an internal server error after 100 attempts

File: pkg/storage/filepath/fs.go (hunk 6 above)
Comment (milas) [reply pkg/storage/filepath/fs.go]: 😅  removed in a3b384c
